components/region_cache_memory_engine/src/memory_controller.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+use crossbeam::utils::CachePadded;
+
+use crate::{engine::SkiplistEngine, write_batch::NODE_OVERHEAD_SIZE_EXPECTATION};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum MemoryUsage {
+    NormalUsage(usize),
+    SoftLimitReached(usize),
+    // usize here means the current memory usage and it's the usize in it adding with the memory
+    // acquiring exceeds the hard limit
+    HardLimitReached(usize),
+}
+
+#[derive(Debug)]
+struct HotData {
+    // Allocated memory for keys and values (node overhead is not included)
+    allocated: AtomicUsize,
+    // The number of writes that are buffered but not yet written.
+    pending_node_count: AtomicUsize,
+}
+
+/// MemoryController is used to control the memory usage of the region cache
+/// memory engine. The memory consumption is calculated by the allocated memory
+/// for keys and values, and the overhead of the skiplist where the overhead is
+/// estimated by using the node count of the skiplist times the
+/// `NODE_OVERHEAD_SIZE_EXPECTATION`.
+#[derive(Debug)]
+pub struct MemoryController {
+    hot_data: CachePadded<HotData>,
+    soft_limit_threshold: usize,
+    hard_limit_threshold: usize,
+    memory_checking: AtomicBool,
+    skiplist_engine: SkiplistEngine,
+}
+
+impl MemoryController {
+    pub fn new(
+        soft_limit_threshold: usize,
+        hard_limit_threshold: usize,
+        skiplist_engine: SkiplistEngine,
+    ) -> Self {
+        Self {
+            hot_data: CachePadded::new(HotData {
+                allocated: AtomicUsize::new(0),
+                pending_node_count: AtomicUsize::new(0),
+            }),
+            soft_limit_threshold,
+            hard_limit_threshold,
+            memory_checking: AtomicBool::new(false),
+            skiplist_engine,
+        }
+    }
+
+    pub(crate) fn acquire(&self, n: usize) -> MemoryUsage {
+        let pending_count = self
+            .hot_data
+            .pending_node_count
+            .fetch_add(1, Ordering::Relaxed)
+            + 1;
+        let node_count = self.skiplist_engine.node_count();
+
+        let mem_usage = self.hot_data.allocated.fetch_add(n, Ordering::Relaxed)
+            + n
+            + (node_count + pending_count) * NODE_OVERHEAD_SIZE_EXPECTATION;
+        if mem_usage >= self.hard_limit_threshold {
+            self.hot_data.allocated.fetch_sub(n, Ordering::Relaxed);
+            self.hot_data
+                .pending_node_count
+                .fetch_sub(1, Ordering::Relaxed);
+            return MemoryUsage::HardLimitReached(mem_usage - n - NODE_OVERHEAD_SIZE_EXPECTATION);
+        }
+
+        if mem_usage >= self.soft_limit_threshold {
+            return MemoryUsage::SoftLimitReached(mem_usage);
+        }
+
+        MemoryUsage::NormalUsage(mem_usage)
+    }
+
+    pub(crate) fn on_node_written(&self, counts: usize) {
```

Comment:
I think generally it's not needed for having pending_node_count.
We can use self.skiplist_engine.node_count() directly, as the difference is neglectable.  
This can save some CPU usage and also make the code a bit simpler.